File: mobile/ios/QwenMessagingSDK.swift
//
//  QwenMessagingSDK.swift
//  iOS SDK for Qwen Messaging Agent
//
//  Complete SDK for iOS applications with authentication, caching, and error handling
//

import Foundation
import Combine

// MARK: - Models

public struct ChatMessage: Codable {
    public let role: String
    public let content: String
    public let timestamp: Date
    
    public init(role: String, content: String, timestamp: Date = Date()) {
        self.role = role
        self.content = content
        self.timestamp = timestamp
    }
}

public struct ChatRequest: Codable {
    public let message: String
    public let conversationId: String?
    public let useCache: Bool
    
    enum CodingKeys: String, CodingKey {
        case message
        case conversationId = "conversation_id"
        case useCache = "use_cache"
    }
}

public struct ChatResponse: Codable {
    public let response: String
    public let conversationId: String
    public let traceId: String?
    public let cached: Bool
    public let durationMs: Double?
    
    enum CodingKeys: String, CodingKey {
        case response
        case conversationId = "conversation_id"
        case traceId = "trace_id"
        case cached
        case durationMs = "duration_ms"
    }
}

public struct AuthToken: Codable {
    public let accessToken: String
    public let refreshToken: String
    public let tokenType: String
    
    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case tokenType = "token_type"
    }
}

// MARK: - SDK Configuration

public struct QwenSDKConfig {
    public let baseURL: String
    public let apiKey: String?
    public let enableCaching: Bool
    public let enableLogging: Bool
    public let timeout: TimeInterval
    
    public init(
        baseURL: String,
        apiKey: String? = nil,
        enableCaching: Bool = true,
        enableLogging: Bool = false,
        timeout: TimeInterval = 30
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.enableCaching = enableCaching
        self.enableLogging = enableLogging
        self.timeout = timeout
    }
}

// MARK: - SDK Error

public enum QwenSDKError: Error, LocalizedError {
    case invalidURL
    case networkError(Error)
    case authenticationFailed
    case invalidResponse
    case serverError(String)
    case rateLimitExceeded
    case timeout
    
    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .authenticationFailed:
            return "Authentication failed"
        case .invalidResponse:
            return "Invalid response from server"
        case .serverError(let message):
            return "Server error: \(message)"
        case .rateLimitExceeded:
            return "Rate limit exceeded"
        case .timeout:
            return "Request timeout"
        }
    }
}

// MARK: - Main SDK

public class QwenMessagingSDK {
    
    private let config: QwenSDKConfig
    private var accessToken: String?
    private var refreshToken: String?
    private let session: URLSession
    private var conversations: [String: [ChatMessage]] = [:]
    private let cache = NSCache<NSString, AnyObject>()
    
    // Combine publishers
    public let messagePublisher = PassthroughSubject<ChatMessage, Never>()
    public let errorPublisher = PassthroughSubject<QwenSDKError, Never>()
    
    public init(config: QwenSDKConfig) {
        self.config = config
        
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = config.timeout
        configuration.timeoutIntervalForResource = config.timeout * 2
        self.session = URLSession(configuration: configuration)
        
        if config.enableLogging {
            print("[QwenSDK] Initialized with base URL: \(config.baseURL)")
        }
    }
    
    // MARK: - Authentication
    
    public func login(username: String, password: String) async throws {
        let url = URL(string: "\(config.baseURL)/auth/token")!
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        let body = "username=\(username)&password=\(password)"
        request.httpBody = body.data(using: .utf8)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw QwenSDKError.invalidResponse
        }
        
        guard httpResponse.statusCode == 200 else {
            throw QwenSDKError.authenticationFailed
        }
        
        let authToken = try JSONDecoder().decode(AuthToken.self, from: data)
        self.accessToken = authToken.accessToken
        self.refreshToken = authToken.refreshToken
        
        if config.enableLogging {
            print("[QwenSDK] Login successful")
        }
    }
    
    public func logout() async throws {
        guard let token = accessToken else { return }
        
        let url = URL(string: "\(config.baseURL)/auth/logout")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        
        _ = try await session.data(for: request)
        
        self.accessToken = nil
        self.refreshToken = nil
        self.conversations.removeAll()
        
        if config.enableLogging {
            print("[QwenSDK] Logged out")
        }
    }
    
    // MARK: - Chat
    
    public func sendMessage(
        _ message: String,
        conversationId: String? = nil,
        useCache: Bool = true
    ) async throws -> ChatResponse {
        
        // Check cache first
        if useCache && config.enableCaching {
            let cacheKey = message.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
            if let cached = cache.object(forKey: cacheKey as NSString) as? ChatResponse {
                if config.enableLogging {
                    print("[QwenSDK] Cache hit for message")
                }
                return cached
            }
        }
        
        // Prepare request
        let url = URL(string: "\(config.baseURL)/chat")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add authentication
        if let token = accessToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if let apiKey = config.apiKey {
            request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        }
        
        // Encode body
        let chatRequest = ChatRequest(
            message: message,
            conversationId: conversationId,
            useCache: useCache
        )
        request.httpBody = try JSONEncoder().encode(chatRequest)
        
        // Send request
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw QwenSDKError.invalidResponse
        }
        
        // Handle errors
        switch httpResponse.statusCode {
        case 200:
            let chatResponse = try JSONDecoder().decode(ChatResponse.self, from: data)
            
            // Cache response
            if useCache && config.enableCaching {
                let cacheKey = message.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
                cache.setObject(chatResponse as AnyObject, forKey: cacheKey as NSString)
            }
            
            // Update conversation history
            let convId = chatResponse.conversationId
            if conversations[convId] == nil {
                conversations[convId] = []
            }
            conversations[convId]?.append(ChatMessage(role: "user", content: message))
            conversations[convId]?.append(ChatMessage(role: "assistant", content: chatResponse.response))
            
            // Publish message
            messagePublisher.send(ChatMessage(role: "assistant", content: chatResponse.response))
            
            if config.enableLogging {
                print("[QwenSDK] Message sent successfully. Cached: \(chatResponse.cached)")
            }
            
            return chatResponse
            
        case 401:
            throw QwenSDKError.authenticationFailed
        case 429:
            throw QwenSDKError.rateLimitExceeded
        case 500...599:
            throw QwenSDKError.serverError("Server error: \(httpResponse.statusCode)")
        default:
            throw QwenSDKError.invalidResponse
        }
    }
    
    // MARK: - Conversation Management
    
    public func getConversationHistory(_ conversationId: String) -> [ChatMessage]? {
        return conversations[conversationId]
    }
    
    public func clearConversation(_ conversationId: String) {
        conversations.removeValue(forKey: conversationId)
    }
    
    public func clearAllConversations() {
        conversations.removeAll()
    }
    
    // MARK: - Cache Management
    
    public func clearCache() {
        cache.removeAllObjects()
        if config.enableLogging {
            print("[QwenSDK] Cache cleared")
        }
    }
}

// MARK: - SwiftUI Integration

#if canImport(SwiftUI)
import SwiftUI

@available(iOS 13.0, *)
public class QwenChatViewModel: ObservableObject {
    @Published public var messages: [ChatMessage] = []
    @Published public var isLoading: Bool = false
    @Published public var errorMessage: String?
    
    private let sdk: QwenMessagingSDK
    private var conversationId: String?
    private var cancellables = Set<AnyCancellable>()
    
    public init(sdk: QwenMessagingSDK) {
        self.sdk = sdk
        
        // Subscribe to message updates
        sdk.messagePublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] message in
                self?.messages.append(message)
            }
            .store(in: &cancellables)
        
        sdk.errorPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] error in
                self?.errorMessage = error.localizedDescription
            }
            .store(in: &cancellables)
    }
    
    public func sendMessage(_ text: String) {
        isLoading = true
        errorMessage = nil
        
        // Add user message immediately
        messages.append(ChatMessage(role: "user", content: text))
        
        Task {
            do {
                let response = try await sdk.sendMessage(
                    text,
                    conversationId: conversationId
                )
                
                // Store conversation ID
                if conversationId == nil {
                    conversationId = response.conversationId
                }
                
                await MainActor.run {
                    isLoading = false
                }
            } catch {
                await MainActor.run {
                    isLoading = false
                    errorMessage = error.localizedDescription
                }
            }
        }
    }
    
    public func clearChat() {
        messages.removeAll()
        conversationId = nil
        errorMessage = nil
    }
}
#endif

File: mobile/android/QwenMessagingSDK.kt
/**
 * QwenMessagingSDK.kt
 * Android SDK for Qwen Messaging Agent
 * 
 * Complete SDK for Android applications with authentication, caching, and error handling
 */

package com.qwen.messaging.sdk

import android.content.Context
import android.util.LruCache
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

// MARK: - Data Models

data class ChatMessage(
    val role: String,
    val content: String,
    val timestamp: Long = System.currentTimeMillis()
)

data class ChatRequest(
    val message: String,
    val conversationId: String? = null,
    val useCache: Boolean = true
)

data class ChatResponse(
    val response: String,
    val conversationId: String,
    val traceId: String? = null,
    val cached: Boolean = false,
    val durationMs: Double? = null
)

data class AuthToken(
    val accessToken: String,
    val refreshToken: String,
    val tokenType: String
)

// MARK: - SDK Configuration

data class QwenSDKConfig(
    val baseURL: String,
    val apiKey: String? = null,
    val enableCaching: Boolean = true,
    val enableLogging: Boolean = false,
    val timeout: Long = 30,
    val cacheSize: Int = 50
)

// MARK: - SDK Error

sealed class QwenSDKError : Exception() {
    object InvalidURL : QwenSDKError()
    data class NetworkError(val error: Throwable) : QwenSDKError()
    object AuthenticationFailed : QwenSDKError()
    object InvalidResponse : QwenSDKError()
    data class ServerError(val message: String) : QwenSDKError()
    object RateLimitExceeded : QwenSDKError()
    object Timeout : QwenSDKError()
    
    override val message: String
        get() = when (this) {
            is InvalidURL -> "Invalid API URL"
            is NetworkError -> "Network error: ${error.message}"
            is AuthenticationFailed -> "Authentication failed"
            is InvalidResponse -> "Invalid response from server"
            is ServerError -> "Server error: $message"
            is RateLimitExceeded -> "Rate limit exceeded"
            is Timeout -> "Request timeout"
        }
}

// MARK: - Main SDK

class QwenMessagingSDK(
    private val config: QwenSDKConfig,
    private val context: Context
) {
    
    private var accessToken: String? = null
    private var refreshToken: String? = null
    
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(config.timeout, TimeUnit.SECONDS)
        .readTimeout(config.timeout, TimeUnit.SECONDS)
        .writeTimeout(config.timeout, TimeUnit.SECONDS)
        .addInterceptor(LoggingInterceptor(config.enableLogging))
        .build()
    
    private val conversations = mutableMapOf<String, MutableList<ChatMessage>>()
    private val cache = LruCache<String, ChatResponse>(config.cacheSize)
    
    // State flows for reactive updates
    private val _messageFlow = MutableStateFlow<ChatMessage?>(null)
    val messageFlow: StateFlow<ChatMessage?> = _messageFlow
    
    private val _errorFlow = MutableStateFlow<QwenSDKError?>(null)
    val errorFlow: StateFlow<QwenSDKError?> = _errorFlow
    
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // MARK: - Authentication
    
    suspend fun login(username: String, password: String) = withContext(Dispatchers.IO) {
        val url = "${config.baseURL}/auth/token"
        
        val formBody = FormBody.Builder()
            .add("username", username)
            .add("password", password)
            .build()
        
        val request = Request.Builder()
            .url(url)
            .post(formBody)
            .build()
        
        try {
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                val json = JSONObject(response.body?.string() ?: "")
                accessToken = json.getString("access_token")
                refreshToken = json.getString("refresh_token")
                
                if (config.enableLogging) {
                    println("[QwenSDK] Login successful")
                }
            } else {
                throw QwenSDKError.AuthenticationFailed
            }
        } catch (e: IOException) {
            throw QwenSDKError.NetworkError(e)
        }
    }
    
    suspend fun logout() = withContext(Dispatchers.IO) {
        val token = accessToken ?: return@withContext
        
        val url = "${config.baseURL}/auth/logout"
        val request = Request.Builder()
            .url(url)
            .post("".toRequestBody())
            .addHeader("Authorization", "Bearer $token")
            .build()
        
        try {
            client.newCall(request).execute()
            accessToken = null
            refreshToken = null
            conversations.clear()
            
            if (config.enableLogging) {
                println("[QwenSDK] Logged out")
            }
        } catch (e: IOException) {
            throw QwenSDKError.NetworkError(e)
        }
    }
    
    // MARK: - Chat
    
    suspend fun sendMessage(
        message: String,
        conversationId: String? = null,
        useCache: Boolean = true
    ): ChatResponse = withContext(Dispatchers.IO) {
        
        // Check cache first
        if (useCache && config.enableCaching) {
            val cacheKey = message.lowercase().trim()
            cache.get(cacheKey)?.let { cached ->
                if (config.enableLogging) {
                    println("[QwenSDK] Cache hit for message")
                }
                return@withContext cached
            }
        }
        
        // Prepare request
        val url = "${config.baseURL}/chat"
        
        val json = JSONObject().apply {
            put("message", message)
            conversationId?.let { put("conversation_id", it) }
            put("use_cache", useCache)
        }
        
        val requestBody = json.toString()
            .toRequestBody("application/json".toMediaType())
        
        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .apply {
                // Add authentication
                accessToken?.let {
                    addHeader("Authorization", "Bearer $it")
                } ?: config.apiKey?.let {
                    addHeader("X-API-Key", it)
                }
            }
            .build()
        
        try {
            val response = client.newCall(request).execute()
            
            when (response.code) {
                200 -> {
                    val responseJson = JSONObject(response.body?.string() ?: "")
                    val chatResponse = ChatResponse(
                        response = responseJson.getString("response"),
                        conversationId = responseJson.getString("conversation_id"),
                        traceId = responseJson.optString("trace_id"),
                        cached = responseJson.optBoolean("cached", false),
                        durationMs = responseJson.optDouble("duration_ms")
                    )
                    
                    // Cache response
                    if (useCache && config.enableCaching) {
                        val cacheKey = message.lowercase().trim()
                        cache.put(cacheKey, chatResponse)
                    }
                    
                    // Update conversation history
                    val convId = chatResponse.conversationId
                    if (!conversations.containsKey(convId)) {
                        conversations[convId] = mutableListOf()
                    }
                    conversations[convId]?.add(ChatMessage("user", message))
                    conversations[convId]?.add(ChatMessage("assistant", chatResponse.response))
                    
                    // Emit message
                    _messageFlow.value = ChatMessage("assistant", chatResponse.response)
                    
                    if (config.enableLogging) {
                        println("[QwenSDK] Message sent. Cached: ${chatResponse.cached}")
                    }
                    
                    return@withContext chatResponse
                }
                401 -> throw QwenSDKError.AuthenticationFailed
                429 -> throw QwenSDKError.RateLimitExceeded
                in 500..599 -> throw QwenSDKError.ServerError("Server error: ${response.code}")
                else -> throw QwenSDKError.InvalidResponse
            }
        } catch (e: IOException) {
            _errorFlow.value = QwenSDKError.NetworkError(e)
            throw QwenSDKError.NetworkError(e)
        }
    }
    
    // MARK: - Conversation Management
    
    fun getConversationHistory(conversationId: String): List<ChatMessage>? {
        return conversations[conversationId]
    }
    
    fun clearConversation(conversationId: String) {
        conversations.remove(conversationId)
    }
    
    fun clearAllConversations() {
        conversations.clear()
    }
    
    // MARK: - Cache Management
    
    fun clearCache() {
        cache.evictAll()
        if (config.enableLogging) {
            println("[QwenSDK] Cache cleared")
        }
    }
    
    fun getCacheSize(): Int {
        return cache.size()
    }
    
    // MARK: - Cleanup
    
    fun cleanup() {
        scope.cancel()
    }
}

// MARK: - Logging Interceptor

private class LoggingInterceptor(private val enabled: Boolean) : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        
        if (enabled) {
            println("[QwenSDK] Request: ${request.method} ${request.url}")
        }
        
        val response = chain.proceed(request)
        
        if (enabled) {
            println("[QwenSDK] Response: ${response.code} (${response.message})")
        }
        
        return response
    }
}

// MARK: - Jetpack Compose Integration

@androidx.compose.runtime.Composable
fun rememberQwenSDK(config: QwenSDKConfig): QwenMessagingSDK {
    val context = androidx.compose.ui.platform.LocalContext.current
    return androidx.compose.runtime.remember { QwenMessagingSDK(config, context) }
}
